base/pmap.jl

```diff
@@ -15,38 +15,81 @@ Note that `f` must be made available to all worker processes; see
+    end
+
+    if batch_size == :auto
+        batches = batchsplit(c, min_batch_count = length(p) * 3)
+    else
+        batches = batchsplit(c, max_batch_size = batch_size)
     end
-    batches = batchsplit(c, min_batch_count = length(p) * 3)
-    return flatten(AsyncGenerator(remote(p, b -> asyncmap(f, b)), batches))
+    return flatten(AsyncGenerator(remote(p, b -> asyncmap(f, b; on_error=on_error)), batches; on_error=on_error))
 end
 
-pgenerate(p::WorkerPool, f, c1, c...) = pgenerate(p, a->f(a...), zip(c1, c...))
+pgenerate(p::WorkerPool, f, c1, c...; kwargs...) = pgenerate(p, a->f(a...), zip(c1, c...); kwargs...)
 
-pgenerate(f, c) = pgenerate(default_worker_pool(), f, c...)
-pgenerate(f, c1, c...) = pgenerate(a->f(a...), zip(c1, c...))
+pgenerate(f, c; kwargs...) = pgenerate(default_worker_pool(), f, c...; kwargs...)
+pgenerate(f, c1, c...; kwargs...) = pgenerate(a->f(a...), zip(c1, c...); kwargs...)
 
 
 """
-    pmap([::WorkerPool], f, c...) -> collection
+    pmap([::WorkerPool], f, c...; distributed=true, batch_size=1, on_error=nothing) -> collection
 
 Transform collection `c` by applying `f` to each element using available
 workers and tasks.
 
 For multiple collection arguments, apply f elementwise.
 
-Note that `err_retry=true` and `err_stop=false` are deprecated,
-use `pmap(retry(f), c)` or `pmap(@catch(f), c)` instead
-(or to retry on a different worker, use `asyncmap(retry(remote(f)), c)`).
-
 Note that `f` must be made available to all worker processes; see
 [Code Availability and Loading Packages](:ref:`Code Availability
 and Loading Packages <man-parallel-computing-code-availability>`)
 for details.
+
+If a worker pool is not specified, all available workers, i.e., the default worker pool
+is used.
+
+By default, `pmap` distributes the computation over all specified workers. To use only the
+local process and distribute over tasks, specifiy `distributed=false`
+
+`pmap` can also use a mix of processes and tasks via the `batch_size` argument. For batch sizes
+greater than 1, the collection is split into multiple batches, which are distributed across
+workers. Each such batch is processed in parallel via tasks in each worker. `batch_size=:auto`
+will automtically calculate a batch size depending on the length of the collection and number
+of workers available.
+
+Any error stops pmap from processing the remainder of the collection. To override this behavior
+you can specify an error handling function via argument `on_error` which takes in a single argument, i.e.,
+the exception. The function can stop the processing by rethrowing the error, or, to continue, return any value
+which is then returned inline with the results to the caller.
 """
-pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+function pmap(p::WorkerPool, f, c...; kwargs...)
+    kwdict = merge(DEFAULT_PMAP_ARGS, AnyDict(kwargs))
+    validate_pmap_kwargs(kwdict, PMAP_KW_NAMES)
+
+    collect(pgenerate(p, f, c...; config=kwdict))
+end
```

Comment:
If deprecated.jl lists the old args and the new args are listed in `pgenerate`, then the defaults are only defined in one place, and the errors for unknown args will work as usual.

deprecated.jl:
```julia
function pmap(f, c...; err_retry=nothing, err_stop=nothing, pids=nothing, kwargs...)
    kwargs = Dict{Symbol, Any}(kwargs)
    ...
    if err_stop == false
        kwargs[:on_error] = e -> e
    end
    ...
    pmap(p, f, c...; kwargs...)
end
```

pmap.jl:
```julia
pmap(p::WorkerPool, f, c...; kwargs...) = collect(pgenerate(p, f, c...; kwargs...))

function pgenerate(p::WorkerPool, f, c;  distributed=true, batch_size=1, on_error=nothing)
    ...
```

Here is a proof of concept typed into the REPL:
```julia
julia> function xx_pmap(f, c...; err_retry=nothing, err_stop=nothing, pids=nothing, kwargs...)
           kwargs = Dict{Symbol, Any}(kwargs)

           if err_stop == false
               kwargs[:on_error] = e -> e
           end
           p = WorkerPool()
           xx_pmap(p, f, c...; kwargs...)
       end
xx_pmap (generic function with 1 method)

julia> xx_pmap(p::WorkerPool, f, c...; kwargs...) = collect(xx_pgenerate(p, f, c...; kwargs...))
xx_pmap (generic function with 2 methods)

julia> function xx_pgenerate(p::WorkerPool, f, c;  distributed=true, batch_size=1, on_error=nothing)
           @show distributed
           @show batch_size
           @show on_error
           map(f,c)
       end
xx_pgenerate (generic function with 1 method)

julia> xx_pmap(x->x*x, [1,2,3])
distributed = true
batch_size = 1
on_error = nothing
3-element Array{Int64,1}:
 1
 4
 9

# Invalid kw arg is rejected:
julia> xx_pmap(x->x*x, [1,2,3], some_option=false)
ERROR: unrecognized keyword argument "some_option"
 in xx_pmap at none:1
 in xx_pmap at none:8

# Deprecated kw arg is accepted:
julia> xx_pmap(x->x*x, [1,2,3], err_retry=true)
distributed = true
batch_size = 1
on_error = nothing
3-element Array{Int64,1}:
 1
 4
 9

# err_stop causes on_error to be set to anonymous function (e -> e):
julia> xx_pmap(x->x*x, [1,2,3], err_stop=false)
distributed = true
batch_size = 1
on_error = (anonymous function)
3-element Array{Int64,1}:
 1
 4
 9

# New kw option is passed though:
julia> xx_pmap(x->x*x, [1,2,3], batch_size=100)
distributed = true
batch_size = 100
on_error = nothing
3-element Array{Int64,1}:
 1
 4
 9
```